src/System.Collections/src/System/Collections/BitArray.cs

```diff
@@ -506,169 +506,206 @@ public void CopyTo(Array array, int index)
-        /// be (n+31)/32, but that would cause overflow. So instead it's 
-        /// rearranged to ((n-1)/32) + 1, with special casing for 0.
+        /// be (n + 31) / 32, but that would cause overflow. So instead it's 
+        /// rearranged to ((n - 1) / 32) + 1.
+        /// Due to sign extension, we don't need to special case for n == 0, if we use
+        /// bitwise operations (since ((n - 1) >> 5) + 1 = 0).
+        /// This doesn't hold true for ((n - 1) / 32) + 1, which equals 1.
         /// 
         /// Usage:
-        /// GetArrayLength(77, BitsPerInt32): returns how many ints must be 
+        /// GetArrayLength(77): returns how many ints must be 
         /// allocated to store 77 bits.
         /// </summary>
         /// <param name="n"></param>
-        /// <param name="div">use a conversion constant, e.g. BytesPerInt32 to get
-        /// how many ints are required to store n bytes</param>
-        /// <returns></returns>
-        private static int GetArrayLength(int n, int div)
+        /// <returns>how many ints are required to store n bytes</returns>
+        private static int GetInt32ArrayLengthFromBitLength(int n)
+        {
+            Debug.Assert(n >= 0);
+            return ((n - 1) >> BitShiftPerInt32) + 1;
+        }
+
+        private static int GetInt32ArrayLengthFromByteLength(int n)
+        {
+            Debug.Assert(n >= 0);
+            // Due to sign extension, we don't need to special case for n == 0, since ((n - 1) >> 2) + 1 = 0
+            // This doesn't hold true for ((n - 1) / 4) + 1, which equals 1.
+            return ((n - 1) >> BitShiftForBytesPerInt32) + 1;
+        }
+
+        private static int GetByteArrayLengthFromBitLength(int n)
         {
-            Debug.Assert(div > 0, "GetArrayLength: div arg must be greater than 0");
-            return n > 0 ? (((n - 1) / div) + 1) : 0;
+            Debug.Assert(n >= 0);
+            // Due to sign extension, we don't need to special case for n == 0, since ((n - 1) >> 3) + 1 = 0
+            // This doesn't hold true for ((n - 1) / 8) + 1, which equals 1.
+            return ((n - 1) >> BitShiftPerByte) + 1;
+        }
+
+        private static int Div32Rem(int number, out int remainder)
+        {
+            uint quotient = (uint)number / BitsPerInt32;
+            remainder = number & (BitsPerInt32 - 1);    // number & 31 == number % 32
+            return (int)quotient;
+        }
+
+        private static int Div4Rem(int number, out int remainder)
+        {
+            uint quotient = (uint)number / BytesPerInt32;
+            remainder = number & (BytesPerInt32 - 1);   // number & 4 == number % 4
+            return (int)quotient;
         }
 
         private class BitArrayEnumeratorSimple : IEnumerator, ICloneable
         {
-            private BitArray bitarray;
-            private int index;
-            private int version;
-            private bool currentElement;
+            private BitArray _bitarray;
```

Comment:
ByteArray is serializable, this might be breaking serialization cc: @ViktorHofer

Reply:
The enumerator is not part of the serialization payload so that change is fine. (tests would complain too)

Reply:
K, just weren't sure :smile:, thanks